File: hotel/backend/src/main/kotlin/loshica/api/hotel/controllers/RoomController.kt
package loshica.api.hotel.controllers

import loshica.api.hotel.annotations.Auth
import loshica.api.hotel.dtos.DeleteDto
import loshica.api.hotel.dtos.RoomDto
import loshica.api.hotel.dtos.RoomPaginationDto
import loshica.api.hotel.dtos.RoomPopulatedDto
import loshica.api.hotel.models.*
import loshica.api.hotel.interfaces.*
import loshica.api.hotel.shared.*
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping(Route.ROOMS)
@CrossOrigin(originPatterns = ["*"])
class RoomController(
    bookingService: IBookingService,
    private val buildingService: IBuildingService,
    commentService: ICommentService,
    optionService: IOptionService,
    private val roomService: IRoomService,
    private val typeService: ITypeService,
    userService: IUserService
) {

    private val destroyer = Destroyer(
        bookingService = bookingService,
        buildingService = buildingService,
        commentService = commentService,
        optionService = optionService,
        roomService = roomService,
        typeService = typeService,
        userService = userService
    )

    @GetMapping
    fun getAll(): List<RoomPopulatedDto> = roomService.getAll().map { it.toPopulatedDto() }

    @GetMapping(Route.PAGINATED)
    fun getWithPagination(
        @RequestParam buildingId: Int?,
        @RequestParam typeId: Int?,
        @RequestParam isFree: Boolean?,
        @RequestParam page: Int?,
        @RequestParam limit: Int?
    ): RoomPaginationDto {
        val requestPage: Int = page ?: 1
        val requestLimit: Int = limit ?: Configuration.ROOM_LIMIT
        val offset: Int = requestPage * requestLimit - requestLimit

        return RoomPaginationDto(
            amount = roomService.getAmount(
                buildingId = buildingId,
                typeId = typeId,
                isFree = isFree
            ),
            rooms = roomService
                .get(
                    buildingId = buildingId,
                    typeId = typeId,
                    isFree = isFree,
                    limit = requestLimit,
                    offset = offset
                )
                .toList().map { it.toPopulatedDto() }
        )
    }

    @GetMapping(Selector.ID)
    fun getOne(@PathVariable id: Int): RoomPopulatedDto = roomService.getOne(id).toPopulatedDto()

    @PostMapping
    fun create(
        @Auth(Role.ADMIN) user: User,
        @RequestBody dto: RoomDto
    ): RoomPopulatedDto {
        val room: Room = roomService.create(
            building = buildingService.getOne(dto.building),
            type = typeService.getOne(dto.type)
        )

        buildingService.addRoom(room)
        return room.toPopulatedDto()
    }

    @PutMapping(Selector.ID)
    fun change(
        @Auth(Role.ADMIN) user: User,
        @RequestBody dto: RoomDto,
        @PathVariable id: Int
    ): RoomPopulatedDto {
        var room: Room = roomService.getOne(id)

        buildingService.removeRoom(room)
        room = roomService.change(
            id = id,
            building = buildingService.getOne(dto.building),
            type = typeService.getOne(dto.type)
        )
        buildingService.addRoom(room)

        return room.toPopulatedDto()
    }

    @DeleteMapping(Selector.ID)
    fun delete(
        @Auth(Role.ADMIN) user: User,
        @PathVariable id: Int
    ): DeleteDto {
        val room: Room = roomService.getOne(id)
        return DeleteDto(id = destroyer.deleteRoom(room))
    }
}

File: hotel-mern/controllers/room.controller.ts
import {
  basketService,
  buildingService,
  orderService,
  reviewService,
  roomService,
} from '../services'
import { ModifiedRequest } from '../shared/types'
import { Response } from 'express'
import { GetRoomsDto } from '../shared/dtos'
import { IQuery } from '../shared/interfaces'
import { OrderPopulated, RoomPopulated } from '../shared/models'
import { errorHandler } from '../shared/decorators'

export default class RoomController {
  async get(req: ModifiedRequest & GetRoomsDto, res: Response) {
    let { _building, _type, page, limit, isFree } = req.query

    if (typeof page === 'string') {
      page = parseInt(page) || 1
    }

    if (typeof limit === 'string') {
      limit = parseInt(limit) || 20
    }

    const offset: number = page * limit - limit
    const query: IQuery = {}

    if (isFree) {
      query.isFree = isFree === 'true'
    }

    if (_building && !_type) query._building = _building
    else if (!_building && _type) query._type = _type
    else if (_building && _type) {
      query._building = _building
      query._type = _type
    }

    const amount: number = await roomService.getAmount(query)
    const rooms: RoomPopulated[] = await roomService.get(query, limit, offset)

    return res.json({ rooms, amount })
  }

  @errorHandler
  async create(req: ModifiedRequest, res: Response) {
    const room: RoomPopulated = await roomService.create(
      req.body._building,
      req.body._type
    )
    await buildingService.addRoom(req.body._building, room._id)
    return res.json(room)
  }

  @errorHandler
  async change(req: ModifiedRequest, res: Response) {
    const { _id, _building, _type } = req.body
    const stock: RoomPopulated = await roomService.getOne(_id)
    const room: RoomPopulated = await roomService.change(_id, _building, _type)

    if (_building !== stock._building._id) {
      await buildingService.removeRoom(_id)
      await buildingService.addRoom(_building, _id)
    }

    return res.json(room)
  }

  @errorHandler
  async delete(req: ModifiedRequest, res: Response) {
    const room: RoomPopulated = await roomService.getOne(req.body._id)
    await roomService.delete(room._id)
    await buildingService.removeRoom(room._id)
    await reviewService.deleteWithRoom(room._id)

    if (room._order) {
      const order: OrderPopulated = await orderService.getOne(room._order)
      await orderService.delete(room._order)
      await basketService.removeOrder(order._id)
    }

    return res.json(room._id)
  }
}
